tests/e2e/upgrade_test.go

```diff
@@ -0,0 +1,140 @@
+		//       Better to check the rollout after the upgrade, it will fail if the upgrade failed
+		_ = rancher.DeployRancherManager(
+			rancherHostname,
+			rancherUpgradeChannel,
+			rancherUpgradeVersion,
+			rancherUpgradeHeadVersion,
+			"None",
+			"None",
+		)
+
+		// Wait for Rancher Manager to be restarted
+		// NOTE: 1st or 2nd rollout command can sporadically fail, so better to use Eventually here
+		Eventually(func() string {
+			status, _ := kubectl.RunWithoutErr(
+				"rollout",
+				"--namespace", "cattle-system",
+				"status", "deployment/rancher",
+			)
+			return status
+		}, tools.SetTimeout(4*time.Minute), 30*time.Second).Should(ContainSubstring("successfully rolled out"))
+
+		// Check that all Rancher Manager pods are running
+		Eventually(func() error {
+			checkList := [][]string{
+				{"cattle-system", "app=rancher"},
+				{"cattle-fleet-local-system", "app=fleet-agent"},
+				{"cattle-system", "app=rancher-webhook"},
+			}
+			return rancher.CheckPod(k, checkList)
+		}, tools.SetTimeout(3*time.Minute), 10*time.Second).Should(Not(HaveOccurred()))
+
+		// Check that all pods are using the same version
+		Eventually(func() int {
+			out, _ := kubectl.RunWithoutErr(getImageVersion...)
+			return len(strings.Fields(out))
+		}, tools.SetTimeout(3*time.Minute), 10*time.Second).Should(Equal(1))
+
+		// Get after-upgrade Rancher Manager version
+		// and check that it's different to the before-upgrade version
+		versionAfterUpgrade, err := kubectl.RunWithoutErr(getImageVersion...)
+		Expect(err).To(Not(HaveOccurred()))
+		Expect(versionAfterUpgrade).To(Not(Equal(versionBeforeUpgrade)))
+
+		// Function to check if all Fleet pods are updated and running the new version
+		isFleetControllerUpgradeComplete := func() bool {
+			// Check the rollout status of Fleet pods to ensure they are updated
+			rolloutStatus, err := kubectl.RunWithoutErr(
+				"rollout",
+				"--namespace", "cattle-fleet-system",
+				"status", "deployment/fleet-controller",
+			)
+			if err != nil {
+					return false
+			}
+
+			// Check if the rollout has completed successfully
+			return strings.Contains(rolloutStatus, `deployment "fleet-controller" successfully rolled out`)
+		}
+
+		// Wait for the upgrade to complete by checking if the Fleet rollout is complete
+		Eventually(isFleetControllerUpgradeComplete, tools.SetTimeout(10*time.Minute), 20*time.Second).Should(BeTrue())
+
+		// Get after-upgrade Fleet version
+		// and check that it's different to the before-upgrade version
+		Eventually(func() int {
+			fleetCmdOut, _ := kubectl.RunWithoutErr(getFleetImageVersion...)
+			return len(strings.Fields(fleetCmdOut))
+		}, tools.SetTimeout(5*time.Minute), 10*time.Second).Should(Equal(3))
+
+		// Get Fleet version after upgrade
+		// and check that it's different to the version before upgrade
+		Eventually(func() string {
+			fleetVersionAfterUpgrade, err := kubectl.RunWithoutErr(getFleetImageVersion...)
```

Comment:
Why run this command again? It seems that this block and the previous one do different things, despite using the same command (and following similar comments, comparing L124-125 with L131-132).

The two could be combined into a single block, with a function taking a `Gomega` argument, as can be seen for instance [here](https://onsi.github.io/gomega/#category-3-making-assertions-eminem-the-function-passed-into-codeeventuallycode).

Reply:
After having discussion with you, I have updated the code. I will run the job and make sure it is working.

Reply:
Thanks for the suggestion @weyfonk, It did the iteration much earlier.